extensions/amp-dynamic-css-classes/0.1/amp-dynamic-css-classes.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {parseUrl} from '../../../src/url';
+import {viewerFor} from '../../../src/viewer';
+
+/**
+ * Returns an array of referrers which vary in level of subdomain specificity.
+ *
+ * @param {string} referrer
+ * @return {Array<string>}
+ * @private Visible for testing only!
+ */
+export function referrers_(referrer) {
+  referrer = parseUrl(referrer).hostname;
+  // Split all dots except for the last one.
+  const domains = referrer.split(/\.(?=.+\.)/);
+  let domainBase;
+
+  return domains.reduceRight((referrers, domain) => {
+    if (domainBase) {
+      domain += '.' + domainBase;
+    }
+    domainBase = domain;
+    referrers.push(domain);
+    return referrers;
+  }, []);
+}
+
+/**
+ * Adds CSS classes onto the HTML element.
+ * @param {!window} win
+ * @param {!Array<string>} classes
+ */
+export function addDynamicCssClasses(win, classes) {
+  const documentElement = win.document.documentElement;
+  const classList = documentElement.classList;
+
+  for (let i = 0; i < classes.length; i++) {
+    classList.add(classes[i]);
+  }
+}
+
+
+/**
+ * Adds dynamic css classes onto the HTML element.
+ * The dynamic classes are generated by:
+ * - referrer, with a separate class for each level of subdomain specificity.
+ * - whether or not the document is inside a viewer.
+ */
+const viewer = viewerFor(window);
+const classes = referrers_(window.document.referrer).map(referrer => {
```

Comment:
It's best to wrap this whole thing in a single function. E.g. `setReferrerClasses(window)`. This way we can remove `export` from `addDynamicCssClasses` - seems a bit too granular for exporting.